### ISSUE TITLE
Attachment support?

### ISSUE THREAD

**Issue by gregorybleiker** [nolanlawson/ember-pouch#37]
As far as I can make out ember-pouch does not support attachments, is that correct? I saw that relational-pouch does support them, are there any plans to implement them in ember-pouch?

**Comment by backspace** [comment]
I’m interested in this too. I was planning on discussing an implementation, but having #32 complete would be preferable before writing anything.

How would you see the interface working?

**Comment by gregorybleiker** [comment]
I would have thought it would work similar to the way ember-couchdb-kit does it
```javascript
App.Issue = DS.Model.extend({
  text: DS.attr('string'),
  type: DS.attr('string', {defaultValue: 'issue'}),
  attachments: DS.hasMany('attachment', {async: true})
});

App.Attachment = DS.Model.extend({
  content_type: DS.attr('string'),
  length: DS.attr('number'),
  file_name: DS.attr('string'),
  db: DS.attr('string', {defaultValue: 'boards'})
});
```
and then something like
```javascript
     var attachment = self.get('store').createRecord('attachment', params);
      attachment.save().then(function() {
        model.get('attachments').pushObject(attachment);
        model.reload();
```

Naively, I tried something like this (before I know attachments weren't supported):
```javascript
            var theModel = this.controller.model;
            var count = 0;
            var file = files[count];
            var attachmentId = "%@/%@".fmt(theModel.id, file.name);

            var params = {
                doc_id: theModel.id,
                model_name: Page,
                rev: theModel._data.rev,
                id: attachmentId,
                file: file,
                content_type: file.type,
                length: file.size,
                file_name: file.name
            };

            var attachment = this.get('store').createRecord('attachment', params);
            theModel._data.attachments.push(attachment);
            theModel.save();
```

**Comment by nolanlawson** [comment]
I think this is doable, since `relational-pouch` supports attachments, but my hunc is that we'll need to modify [these lines](https://github.com/nolanlawson/ember-pouch/blob/master/lib/pouchdb-adapter.js#L85-L99) to move attachments from the Ember record to the Pouch document. There may also be other conversion logic we need to write to convert from Ember records to Blobs/base64.

I would be enthusiastic to see a pull request to do this, though! :)

**Comment by gregorybleiker** [comment]
It might also be possible to inject an "attachments" property into Pouch models, but I don't know how one would figure out which models would be models for pouch adapters...

**Comment by gregorybleiker** [comment]
OK, I had an other think about injecting the attachment property, after some googeling it doesn't seem to be a good idea... Do you think it would be acceptable to have a new base class for ember pouch models, so you would have to inherit from PouchModel instead DS.Model? That would imply that all users of the current library would have to change their model code...

**Comment by nolanlawson** [comment]
I think it should probably be handled the same as `rev`, so yeah, you would need to add it to your models in order to use it.

**Comment by gregorybleiker** [comment]
I'm actually still working on the file attachment feature and have a question (might be relational-pouch or ember-pouch, not sure): When storing the model to the database, it seems to me that ember-pouch (or relational-pouch) wraps all properties into a "data" database field rather then having the properties to be database fields. The reason this bothers me is that the attachments property of the model will then have to be handled differently than all the other properties of the model... Any input on this? Am I making myself clear at all?

**Comment by nolanlawson** [comment]
Yep, that's the idea. I put everything on a `data` member just in case there were weird local variables that might cause problems.

So yes, you would need to add a rule for `obj.attachments`. And I'd be happy to accept your pull request for the feature. :)

**Comment by stevebest** [comment]
@gregorybleiker, in case you're still interested, I made this `Transform` for attachments:

```javascript
// app/transforms/attachments.js
import Ember from 'ember';
import DS from 'ember-data';

export default DS.Transform.extend({
  deserialize: function(serialized) {
    return Ember.keys(Ember.A(serialized)).map(function (attachmentName) {
      return {
        name: attachmentName,
        contentType: serialized[attachmentName]['content_type'],
        data: serialized[attachmentName]
      };
    });
  },

  serialize: function(deserialized) {
    if (typeof deserialized === 'undefined') { return; }
    return deserialized.reduce(function (acc, attachment) {
      acc[attachment.name] = attachment.data;
      return acc;
    }, {});
  }
});
```

You can then declare `attachments`-typed attribute on your model:

```javascript
// app/models/foo.js
// .... skip the boring parts
{
  attachments: DS.attr('attachments'),
}
```

And use it as if it was an array of `{ name, contentType, data }` objects:

```javascript
myModel.get('attachments').addObject({ name: 'image.png', data: data });
```

```handlebars
{{#each model.attachments as |attachment|}}
   {{attachment.name}}
{{/each}}
```

**Comment by rsutphin** [comment]
Attachments are being addressed in #97. It would be great if anyone who's interested in ember-pouch and attachments could comment on the proposed solution there.
